### DISCUSSION TITLE
How to get clojurephant/jovial to call `clojure.test/report :end-test-var`?

### DISCUSSION THREAD

**Discussion by ngrewe** [clojurephant/clojurephant#214] [Q&A]
Hi folks,

I'm porting an older project that previously relied on the [clojuresque gradle plugin](https://github.com/clojuresque/base) to clojurephant, and I'm running into a bit of a problem with the JUnit integration. My test suite defines a [custom `report` hook for `:end-test-var`](https://github.com/ngrewe/tqme/blob/fc3e313917a76d45248ac7b5d4998c75be5ec332/src/test/clojure/tqc/cq.clj#L27-L28). When I execute the test suite in IntelliJ using Cursive, that method instead gets called as expected, but when executing `./gradlew test` (using the Jovial integration), it isn't called. Can anybody point me towards why that might be happening and how to fix it? I've already tried using `(defmethod dev.clojurephant.jovial.engine.clojure-test/jovial-report :end-test-var …)` instead of `(defmethod clojure.test/report :end-test-var …)`, but to no avail.

Thanks,

Niels

**Comment by ajoberstar** [comment]
I'd have to look back at Clojure's test runner to see when that function usually gets called. It could be that jovial needs to explicitly call it. However, I'm also curious if you have more context on the use case. Is this a reporting-related use case where you want something custom printed somewhere? Or is it some kind of cleanup task?

**Reply by ngrewe** [reply]
> Is this a reporting-related use case where you want something custom printed somewhere? Or is it some kind of cleanup task?

It's a true reporting thing. It's writing out intermediate artefacts that are used to perform the tests so that they can be inspected manually. And since you're probably hinting at fixtures (‘cleanup’): I've tried that as well, but I can't seem to get the name of a test from a fixture…

**Reply by ajoberstar** [reply]
Looks like (at a minimum) we'll need to replicate how [clojure.test](https://github.com/clojure/clojure/blob/e3520c07e8b21dbf5a91fa14b7114b90dc1e89c6/src/clj/clojure/test.clj#L708-L721) calls begin/end around testing the var. I think we'd just need to add it in [here](https://github.com/clojurephant/jovial/blob/main/src/main/clojure/dev/clojurephant/jovial/engine/clojure_test.clj#L32-L36).

Open to suggestions on if/how we make this extensible for end users. The `(defmethod dev.clojurephant.jovial.engine.clojure-test/jovial-report :end-test-var …)` that you tried is the most straightforward, but I hadn't originally intended to have a public API surface to jovial, because ideally running your test via jovial would not impact what you write in your tests. But if there's not a better way to do it transparently, that could still be an option.

**Reply by ajoberstar** [reply]
It would probably be worth adding begin/end ns reporting in the NamespaceDescriptor as well to complete the set.

**Reply by ajoberstar** [reply]
I opened this issue to track a resolution: https://github.com/clojurephant/jovial/issues/23
